File: 062/cpp/main.cpp
class Solution {
public:
    int uniquePaths(int m, int n, map<tuple<int, int>, int> &cache) {
        if (m == 1) {
            return 1;
        }
        if (n == 1) {
            return 1;
        }

        tuple<int, int> xy = {m, n};
        auto p = cache.find(xy);
        if (p != cache.end()) {
            return p->second;
        }

        int k = uniquePaths(m - 1, n, cache) + uniquePaths(m, n - 1, cache);
        cache[{m, n}] = k;
        return k;
    }

    int uniquePaths(int m, int n) {
        map<tuple<int, int>, int> cache;
        return uniquePaths(m, n, cache);
    }
};

File: 062/rust/main.rs
impl Solution {
    pub fn unique_paths(m: i32, n: i32) -> i32 {
        let m = m as usize;
        let n = n as usize;
        let mut cache = vec![1; n];
        for i in 1..m {
            for j in 1..n {
                cache[j] += cache[j - 1]
            }
        }
        return cache[n - 1];
    }
}
